File: qasdk/src/main/java/co/astrnt/qasdk/repository/SectionRepository.java
package co.astrnt.qasdk.repository;

import java.util.HashMap;

import co.astrnt.qasdk.core.AstronautApi;
import co.astrnt.qasdk.core.MyObserver;
import co.astrnt.qasdk.dao.BaseApiDao;
import co.astrnt.qasdk.dao.InterviewApiDao;
import co.astrnt.qasdk.dao.InterviewStartApiDao;
import co.astrnt.qasdk.dao.LogDao;
import co.astrnt.qasdk.dao.QuestionApiDao;
import co.astrnt.qasdk.dao.SectionApiDao;
import co.astrnt.qasdk.dao.SummarySectionApiDao;
import co.astrnt.qasdk.type.ElapsedTime;
import co.astrnt.qasdk.type.ElapsedTimeType;
import co.astrnt.qasdk.utils.LogUtil;
import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Created by deni rohimat on 25/05/18.
 */
public class SectionRepository extends BaseRepository {

    public SectionRepository(AstronautApi astronautApi) {
        super(astronautApi);
    }

    public Observable<InterviewStartApiDao> startSection(SectionApiDao sectionApiDao) {
        InterviewApiDao interviewApiDao = astrntSDK.getCurrentInterview();

        HashMap<String, String> map = new HashMap<>();
        map.put("interview_code", interviewApiDao.getInterviewCode());
        map.put("candidate_id", String.valueOf(interviewApiDao.getCandidate().getId()));
        map.put("section_id", String.valueOf(sectionApiDao.getId()));
        String token = interviewApiDao.getToken();

        if (!astrntSDK.isSelfPace()) {
            updateElapsedTime(ElapsedTimeType.PREPARATION, sectionApiDao.getId());
        }

        LogUtil.addNewLog(interviewApiDao.getInterviewCode(),
                new LogDao("Hit API (/section/start)",
                        "Start Section, number" + (astrntSDK.getSectionIndex() + 1) +
                                ", sectionId = " + sectionApiDao.getId()
                )

        );
        astrntSDK.saveLastApiCall("(/section/start)");

        astrntSDK.updateSectionOnGoing(sectionApiDao, true);
        astrntSDK.setContinueInterview(true);
        return mAstronautApi.getApiService().startSection(token, map);
    }

    public Observable<BaseApiDao> finishSection(SectionApiDao sectionApiDao) {
        InterviewApiDao interviewApiDao = astrntSDK.getCurrentInterview();

        HashMap<String, String> map = new HashMap<>();
        map.put("interview_code", interviewApiDao.getInterviewCode());
        map.put("candidate_id", String.valueOf(interviewApiDao.getCandidate().getId()));
        map.put("section_id", String.valueOf(sectionApiDao.getId()));
        String token = interviewApiDao.getToken();

        if (!astrntSDK.isSelfPace()) {
            updateElapsedTime(ElapsedTimeType.SECTION, sectionApiDao.getId());
        }

        LogUtil.addNewLog(interviewApiDao.getInterviewCode(),
                new LogDao("Hit API (/section/stop)",
                        "Finish Section, number " + (astrntSDK.getSectionIndex() + 1) +
                                ", sectionId = " + sectionApiDao.getId()
                )

        );
        astrntSDK.saveLastApiCall("(/section/stop)");

        return mAstronautApi.getApiService().stopSection(token, map);
    }

    public Observable<SummarySectionApiDao> summarySection() {
        InterviewApiDao interviewApiDao = astrntSDK.getCurrentInterview();

        HashMap<String, String> map = new HashMap<>();
        map.put("interview_code", interviewApiDao.getInterviewCode());
        String token = interviewApiDao.getToken();

        return mAstronautApi.getApiService().summarySection(token, map);
    }

    private void updateElapsedTime(@ElapsedTime String type, long refId) {
        InterviewApiDao interviewApiDao = astrntSDK.getCurrentInterview();
        final String interviewCode = astrntSDK.getInterviewCode();

        HashMap<String, String> map = new HashMap<>();
        map.put("interview_code", interviewApiDao.getInterviewCode());
        map.put("type", type);
        map.put("ref_id", String.valueOf(refId));

        String token = interviewApiDao.getToken();

        LogUtil.addNewLog(interviewCode,
                new LogDao("Hit API (/interview/update/elapsedTime)",
                        "Update Elapsed Time Section, type = " + type +
                                ", number " + (astrntSDK.getSectionIndex() + 1)
                                + ", refId = " + refId

                )
        );

        astrntSDK.saveLastApiCall("(/interview/update/elapsedTime)");

        mAstronautApi.getApiService().updateElapsedTime(token, map)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .subscribe(new MyObserver<BaseApiDao>() {

                    @Override
                    public void onSubscribe(Disposable d) {
                    }

                    @Override
                    public void onApiResultCompleted() {
                    }

                    @Override
                    public void onApiResultError(String title, String message, String code) {
                        Timber.e(message);

                        LogUtil.addNewLog(interviewCode,
                                new LogDao("Hit API (Elapsed Time Section)",
                                        "Error " + message
                                )
                        );

                        astrntSDK.saveLastApiCall("(Elapsed Time Section)");
                    }

                    @Override
                    public void onApiResultOk(BaseApiDao apiDao) {
                        Timber.d(apiDao.getMessage());
                    }
                });
    }

    public Observable<BaseApiDao> addLastSeen(int questionId) {
        InterviewApiDao interviewApiDao = astrntSDK.getCurrentInterview();
        String token = interviewApiDao.getToken();

        HashMap<String, String> map = new HashMap<>();
        map.put("interview_code", interviewApiDao.getInterviewCode());
        map.put("question_id", String.valueOf(questionId));

        LogUtil.addNewLog(interviewApiDao.getInterviewCode(),
                new LogDao("Hit API (/question/last_seen)",
                        "Add Last Seen, number " + (astrntSDK.getQuestionIndex() + 1) +
                                ", questionId = " + questionId
                )
        );
        astrntSDK.saveLastApiCall("(/question/last_seen)");

        return mAstronautApi.getApiService().addLastSeen(token, map);
    }

}

File: qasdk/src/main/java/co/astrnt/qasdk/repository/SectionRepository.kt
package co.astrnt.qasdk.repository

import co.astrnt.qasdk.core.AstronautApi
import co.astrnt.qasdk.core.MyObserver
import co.astrnt.qasdk.dao.*
import co.astrnt.qasdk.type.ElapsedTime
import co.astrnt.qasdk.type.ElapsedTimeType
import co.astrnt.qasdk.utils.LogUtil.addNewLog
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import java.util.*

class SectionRepository(astronautApi: AstronautApi) : BaseRepository(astronautApi) {
    fun startSection(sectionApiDao: SectionApiDao): Observable<InterviewStartApiDao> {
        val interviewApiDao = astrntSDK.currentInterview
        val map = HashMap<String, String?>()
        map["interview_code"] = interviewApiDao.interviewCode
        map["candidate_id"] = interviewApiDao.candidate!!.id.toString()
        map["section_id"] = sectionApiDao.id.toString()
        val token = interviewApiDao.token
        if (!astrntSDK.isSelfPace) {
            updateElapsedTime(ElapsedTimeType.PREPARATION, sectionApiDao.id)
        }
        addNewLog(interviewApiDao.interviewCode,
                LogDao("Hit API (/section/start)",
                        "Start Section, number" + (astrntSDK.sectionIndex + 1) +
                                ", sectionId = " + sectionApiDao.id
                )
        )
        astrntSDK.saveLastApiCall("(/section/start)")
        astrntSDK.updateSectionOnGoing(sectionApiDao, true)
        astrntSDK.isContinueInterview = true
        return mAstronautApi.apiService.startSection((token)!!, map)
    }

    fun finishSection(sectionApiDao: SectionApiDao): Observable<BaseApiDao> {
        val interviewApiDao = astrntSDK.currentInterview
        val map = HashMap<String, String?>()
        map["interview_code"] = interviewApiDao.interviewCode
        map["candidate_id"] = interviewApiDao.candidate!!.id.toString()
        map["section_id"] = sectionApiDao.id.toString()
        val token = interviewApiDao.token
        if (!astrntSDK.isSelfPace) {
            updateElapsedTime(ElapsedTimeType.SECTION, sectionApiDao.id)
        }
        addNewLog(interviewApiDao.interviewCode,
                LogDao("Hit API (/section/stop)",
                        ("Finish Section, number " + (astrntSDK.sectionIndex + 1) +
                                ", sectionId = " + sectionApiDao.id)
                )
        )
        astrntSDK.saveLastApiCall("(/section/stop)")
        return mAstronautApi.apiService.stopSection((token)!!, map)
    }

    fun summarySection(): Observable<SummarySectionApiDao> {
        val interviewApiDao = astrntSDK.currentInterview
        val map = HashMap<String, String?>()
        map["interview_code"] = interviewApiDao.interviewCode
        val token = interviewApiDao.token
        return mAstronautApi.apiService.summarySection((token)!!, map)
    }

    private fun updateElapsedTime(@ElapsedTime type: String, refId: Long) {
        val interviewApiDao = astrntSDK.currentInterview
        val interviewCode = astrntSDK.interviewCode
        val map = HashMap<String, String?>()
        map["interview_code"] = interviewApiDao.interviewCode
        map["type"] = type
        map["ref_id"] = refId.toString()
        val token = interviewApiDao.token
        addNewLog(interviewCode,
                LogDao("Hit API (/interview/update/elapsedTime)",
                        ("Update Elapsed Time Section, type = " + type +
                                ", number " + (astrntSDK.sectionIndex + 1)
                                + ", refId = " + refId)
                )
        )
        astrntSDK.saveLastApiCall("(/interview/update/elapsedTime)")
        mAstronautApi.apiService.updateElapsedTime((token)!!, map)
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .subscribe(object : MyObserver<BaseApiDao>() {
                    override fun onSubscribe(d: Disposable) {}
                    override fun onApiResultCompleted() {}
                    override fun onApiResultError(title: String?, message: String?, code: String?) {
                        Timber.e(message)
                        addNewLog(interviewCode,
                                LogDao("Hit API (Elapsed Time Section)",
                                        "Error $message"
                                )
                        )
                        astrntSDK.saveLastApiCall("(Elapsed Time Section)")
                    }

                    override fun onApiResultOk(t: BaseApiDao) {
                        Timber.d(t.message)
                    }
                })
    }

    fun addLastSeen(questionId: Int): Observable<BaseApiDao> {
        val interviewApiDao = astrntSDK.currentInterview
        val token = interviewApiDao.token
        val map = HashMap<String, String?>()
        map["interview_code"] = interviewApiDao.interviewCode
        map["question_id"] = questionId.toString()
        addNewLog(interviewApiDao.interviewCode,
                LogDao("Hit API (/question/last_seen)",
                        ("Add Last Seen, number " + (astrntSDK.questionIndex + 1) +
                                ", questionId = " + questionId)
                )
        )
        astrntSDK.saveLastApiCall("(/question/last_seen)")
        return mAstronautApi.apiService.addLastSeen((token)!!, map)
    }
}